contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,591 @@
+    else if (compressSlow && ctx->compressionLevel > 1) {
+        DEBUGLOG(2, "decreasing compression level %u\n", ctx->compressionLevel);
+        ctx->compressionLevel--;
+        reset = 1;
+    }
+    if (reset) {
+        ctx->stats.readyCounter = 0;
+        ctx->stats.writeCounter = 0;
+        ctx->stats.compressedCounter = 0;
+    }
+    return ctx->compressionLevel;
+}
+
+static void* compressionThread(void* arg)
+{
+    adaptCCtx* ctx = (adaptCCtx*)arg;
+    unsigned currJob = 0;
+    for ( ; ; ) {
+        unsigned const currJobIndex = currJob % ctx->numJobs;
+        jobDescription* job = &ctx->jobs[currJobIndex];
+        DEBUGLOG(2, "compressionThread(): waiting on job ready\n");
+        pthread_mutex_lock(&ctx->jobReady_mutex);
+        while(currJob + 1 > ctx->jobReadyID) {
+            ctx->stats.waitReady++;
+            ctx->stats.readyCounter++;
+            DEBUGLOG(2, "waiting on job ready, nextJob: %u\n", currJob);
+            pthread_cond_wait(&ctx->jobReady_cond, &ctx->jobReady_mutex);
+        }
+        pthread_mutex_unlock(&ctx->jobReady_mutex);
+        DEBUGLOG(2, "compressionThread(): continuing after job ready\n");
+        /* compress the data */
+        {
+            unsigned const cLevel = adaptCompressionLevel(ctx);
+            DEBUGLOG(2, "cLevel used: %u\n", cLevel);
+            size_t const compressedSize = ZSTD_compressCCtx(ctx->cctx, job->dst.start, job->dst.size, job->src.start, job->src.size, cLevel);
+            if (ZSTD_isError(compressedSize)) {
+                ctx->threadError = 1;
+                DISPLAY("Error: something went wrong during compression: %s\n", ZSTD_getErrorName(compressedSize));
+                return arg;
+            }
+            job->compressedSize = compressedSize;
+        }
+        pthread_mutex_lock(&ctx->jobCompressed_mutex);
+        ctx->jobCompressedID++;
+        DEBUGLOG(2, "signaling for job %u\n", currJob);
+        pthread_cond_signal(&ctx->jobCompressed_cond);
+        pthread_mutex_unlock(&ctx->jobCompressed_mutex);
+        DEBUGLOG(2, "finished job compression %u\n", currJob);
+        currJob++;
+        if (currJob >= ctx->lastJobID || ctx->threadError) {
+            /* finished compressing all jobs */
+            DEBUGLOG(2, "all jobs finished compressing\n");
+            break;
+        }
+    }
+    return arg;
+}
+
+static void displayProgress(unsigned jobDoneID, unsigned cLevel, unsigned last)
+{
+    if (!g_useProgressBar) return;
+    UTIL_time_t currTime;
+    UTIL_getTime(&currTime);
+    double const timeElapsed = (double)(UTIL_getSpanTimeMicro(g_ticksPerSecond, g_startTime, currTime) / 1000.0);
+    double const sizeMB = (double)g_streamedSize / (1 << 20);
+    double const avgCompRate = sizeMB * 1000 / timeElapsed;
+    fprintf(stdout, "\r| %4u jobs completed | Current Compresion Level: %2u | Time Elapsed: %5.0f ms | Data Size: %7.1f MB | Avg Compression Rate: %6.2f MB/s |", jobDoneID, cLevel, timeElapsed, sizeMB, avgCompRate);
```

Comment:
Yes,
always write accessory notifications to `stderr`,
`stdout` is here reserved for the compressed data itself.